lib/dom.js

```diff
@@ -4,6 +4,117 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
+/* Crypto */
+
+// This is a helper type to simplify the specification, it isn't an interface
+// and there are no objects implementing it.
+// https://developer.mozilla.org/en-US/docs/Web/API/BufferSource
+type BufferSource = ArrayBuffer | $ArrayBufferView;
+
+type DOMString = string;
+
+declare interface Crypto extends RandomSource {
+  subtle: SubtleCrypto; // readonly
+}
+
+type CryptoKey$Type =
+  | 'secret'
+  | 'public'
+  | 'private';
+type CryptoKey$Usages =
+  | 'encrypt'
+  | 'decrypt'
+  | 'sign'
+  | 'verify'
+  | 'deriveKey'
+  | 'deriveBits'
+  | 'wrapKey'
+  | 'unwrapKey';
+declare interface CryptoKey {
+    algorithm: Object; // readonly
+    extractable: Boolean; // readonly
+    type: CryptoKey$Type; // readonly
+    usages: CryptoKey$Usages[]; // readonly
+}
+
+// TODO: CryptoKeyPair
+declare interface CryptoKeyPair {
+
+}
+
+// TODO: JSONWebKey
+type JSONWebKey = {};
+
+type RandomSource =  {
+  getRandomValues(typedArray: $TypedArray): void;
+};
+
+type SubtleCrypto$Algorithm =
+  | {name: 'AES-CBC', iv: BufferSource}
+  | {name: 'AES-CTR', counter: BufferSource, length: number}
+  | {name: 'AES-GCM', iv: BufferSource, additionalData?: BufferSource, tagLength?: number}
+  | {name: 'RSA-OAEP', label?: string};
+type SubtleCrypto$DerivedKeyAlgorithm = {
+  name: DOMString,
+  length?: 128 | 192 | 256,
+};
+// TODO: fix types for parameters
+type SubtleCrypto$DeriveKeyAlgorithm =
+  | {name: 'ECDH', public: any}
+  | {name: 'DH', public: any}
+  | {name: 'PBKDF2', salt: BufferSource, iterations: number, hash: DOMString}
+  | {name: 'HKDF-CTR', hash: any, label: any, context: any};
+type SubtleCrypto$GenerateKeyAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW'
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA'
+  | 'ECDH'
+  | 'DH';
+type SubtleCrypto$HashAlgo =
+  | 'SHA-1'
+  | 'SHA-256'
+  | 'SHA-384'
+  | 'SHA-512';
+type SubtleCrypto$KeyFormat =
+  | 'raw'
+  | 'pkcs8'
+  | 'spki'
+  | 'jwk';
+type SubtleCrypto$SignAlgo =
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA';
+type SubtleCrypto$WrapAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW';
+declare interface SubtleCrypto {
+  decrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<string>;
```

Comment:
Also definitely `Promise<ArrayBuffer>`, the example uses TextDecoder and I looked up type in Chrome and FF debugger.
```js
(async function test(){

const ptUtf8 = new TextEncoder().encode('my plaintext')

const pwUtf8 = new TextEncoder().encode('my password')
const pwHash = await crypto.subtle.digest('SHA-256', pwUtf8)

const iv = crypto.getRandomValues(new Uint8Array(12))
const alg = { name: 'AES-GCM', iv: iv }
const key = await crypto.subtle.importKey('raw', pwHash, alg, false, ['encrypt'])

const ctBuffer = await crypto.subtle.encrypt(alg, key, ptUtf8)

const key1 = await crypto.subtle.importKey('raw', pwHash, alg, false, ['decrypt'])

const ptBuffer = await crypto.subtle.decrypt(alg, key1, ctBuffer)

const plaintext = new TextDecoder().decode(ptBuffer)
debugger
return plaintext
}())
```

Reply:
`BufferSource` is more appropriate I think, since the input (`data `) is of type `BufferSource`?

Reply:
It doesn't preserve the type in any way though? It's impossible to guess what was used from the ciphertext, so it's a safe bet it will always be an `ArrayBuffer`? :thinking:
It literally operates on the raw bytes.